Cabal/src/Distribution/PackageDescription/Configuration.hs

```diff
@@ -598,3 +599,92 @@ transformAllBuildDependsN f =
   . over (L.packageDescription . L.setupBuildInfo . traverse . L.setupDepends) f
   -- cannot be point-free as normal because of higher rank
   . over (\f' -> L.allCondTrees $ traverseCondTreeC f') f
+
+-- | Flatten a generic package description but try to be a bit more clever with conditionals
+-- in particular, attempt not to introduce suprious module duplciation when merging two elements of a branch.
+-- This is used for trying to avoid spurious duplicate module warnings in sanity checking a package, and
+-- in the future can be extended to try to handle other quirks that may arise.
+flattenPackageDescriptionWithConditionals :: GenericPackageDescription -> PackageDescription
+flattenPackageDescriptionWithConditionals
+  (GenericPackageDescription pkg _ _ mlib0 sub_libs0 flibs0 exes0 tests0 bms0) =
+    pkg { library      = mlib
+        , subLibraries = reverse sub_libs
+        , foreignLibs  = reverse flibs
+        , executables  = reverse exes
+        , testSuites   = reverse tests
+        , benchmarks   = reverse bms
+        }
+  where
+    mlib = f <$> mlib0
+      where f lib = (libFillInDefaults . unwrapWrappedLibrary . fst . ignoreConditions . fmap WrappedLibrary $ lib) { libName = LMainLibName }
+    sub_libs = flattenLib  <$> sub_libs0
+    flibs    = flattenFLib <$> flibs0
+    exes     = flattenExe  <$> exes0
+    tests    = flattenTst  <$> tests0
+    bms      = flattenBm   <$> bms0
+    flattenLib (n, t) = libFillInDefaults $ (unwrapWrappedLibrary . fst . ignoreConditions . fmap WrappedLibrary $ t)
+      { libName = LSubLibName n, libExposed = False }
+    flattenFLib (n, t) = flibFillInDefaults $ (fst $ ignoreConditions t)
+      { foreignLibName = n }
+    flattenExe (n, t) = exeFillInDefaults $ (unwrapWrappedExe . fst . ignoreConditions . fmap WrappedExe $ t)
+      { exeName = n }
+    flattenTst (n, t) = testFillInDefaults $ (unwrapWrappedTest . fst . ignoreConditions . fmap WrappedTest $ t)
+      { testName = n }
+    flattenBm (n, t) = benchFillInDefaults $ (unwrapWrappedBench . fst . ignoreConditions . fmap WrappedBench $ t)
+      { benchmarkName = n }
+
+-- special newtypes for combining with conditions in a way that tries not to introduce module duplication across branches
+newtype WrappedLibrary = WrappedLibrary {unwrapWrappedLibrary :: Library}
+
+instance Semigroup WrappedLibrary where
+   WrappedLibrary a <> WrappedLibrary b = WrappedLibrary $ ab
+                                          {exposedModules = exposed', libBuildInfo = lbi', reexportedModules = reexported', signatures = sigs' }
+    where
+      ab = a <> b
+
+      -- if something exists in both branches of a conditional, only add it once
+      o x y = x ++ filter (not . (`elem` x)) y
```

Comment:
There is `ordNub` in https://hackage.haskell.org/package/Cabal-3.6.0.0/docs/Distribution-Utils-Generic.html#v:ordNub

Reply:
The intent isn't exactly like `ordNub` -- it preserves duplication as long as that duplication exists already in either of the branches. So if `x` or `y` contains duplicates, then `x o y` should also contain them. (And as I write this, I see where it actually fails to have that semantics fully, but that's the intention, at least !).